manifests/init.pp

```diff
@@ -23,13 +26,18 @@
 #   to 'u_puppet_environment'.
 class servicenow_cmdb_integration (
   String $instance,
-  String $user,
-  String $password,
-  String $table             = 'cmdb_ci',
-  String $certname_field    = 'fqdn',
-  String $classes_field     = 'u_puppet_classes',
-  String $environment_field = 'u_puppet_environment',
+  Optional[String] $user        = undef,
+  Optional[String] $password    = undef,
+  Optional[String] $oauth_token = undef,
+  String $table                 = 'cmdb_ci',
+  String $certname_field        = 'fqdn',
+  String $classes_field         = 'u_puppet_classes',
+  String $environment_field     = 'u_puppet_environment',
 ) {
+  if (($password == undef) or ($user == undef)) and ($oauth_token == undef) {
```

Comment:
Ditto https://github.com/puppetlabs/puppetlabs-servicenow_cmdb_integration/pull/25/files#r446016935

Reply:
Also, I think we should just have them specify one set of credentials. If they specify both (so oauth_token, user, _and_ password), then we also fail. Reason is b/c I don't see a good reason to have the "`oauth_token` overrides `user/password`" magic.

Reply:
I think we should give a more specific error message for user/password case b/c I can see a very lazy and very tired admin spending more time than they need to to figure out that a message like "user and password or oauth must be specified" means they forgot the user/password.

So, something like this:

```
if user or password
  if user == undef
    fail("missing password")
  else if password == undef
    fail("missing user")
  end
end
```

Reply:
Note that the credentials validation in the `servicenow.rb` script doesn't have to be as thorough as this -- I think what you've got is fine (so defensive, not thorough, programming is more than enough).